apiserver/common/crossmodel/auth.go

```diff
@@ -27,3 +38,325 @@ func NewBakery(st *state.State) (*bakery.Service, error) {
 		Store:    store,
 	})
 }
+
+const (
+	usernameKey    = "username"
+	offerurlKey    = "offer-url"
+	sourcemodelKey = "source-model-uuid"
+	relationKey    = "relation-key"
+
+	offerPermissionCaveat          = "has-offer-permission"
+	localOfferPermissionExpiryTime = 2 * time.Minute
+)
+
+// AuthContext is used to validate macaroons used to access
+// application offers.
+type AuthContext struct {
+	pool StatePool
+
+	clock                             clock.Clock
+	localOfferThirdPartyBakeryService authentication.BakeryService
+	localOfferBakeryService           authentication.ExpirableStorageBakeryService
+
+	offerAccessEndpoint string
+}
+
+// NewAuthContext creates a new authentication context for checking
+// macaroons used with application offer requests.
+func NewAuthContext(
+	pool StatePool,
+	localOfferThirdPartyBakeryService authentication.BakeryService,
+	localOfferBakeryService authentication.ExpirableStorageBakeryService,
+) (*AuthContext, error) {
+	ctxt := &AuthContext{
+		pool:  pool,
+		clock: clock.WallClock,
+		localOfferBakeryService:           localOfferBakeryService,
+		localOfferThirdPartyBakeryService: localOfferThirdPartyBakeryService,
+	}
+	return ctxt, nil
+}
+
+// AuthContextForClock creates a new authentication context
```

Comment:
do you really need the AuthContext in the method names? maybe just "WithClock" and "WithDischargeURL"?